mlc_llm/transform/smoothquant.py

```diff
@@ -0,0 +1,245 @@
+import tvm
+from tvm import relax
+from tvm.relax.analysis import remove_all_unused
+from tvm.relax.dpl import rewrite_call, is_op, wildcard
+from tvm.relax.expr_functor import PyExprMutator, mutator
+from tvm.script import relax as R
+
+from typing import Dict
+
+
+@mutator
+class Annotator(PyExprMutator):
+    def __init__(self, irmod: tvm.IRModule, mode: str) -> None:
+        super().__init__(irmod)
+        self.mod = irmod
+        self.sm_counter = 0
+        self.new_params = []
+        self.mode = mode
+
+    def transform(self) -> tvm.IRModule:
+        for gv, func in self.mod.functions.items():
+            if not isinstance(func, relax.Function):
+                continue
+            self.sm_counter = 0
+            self.new_params = []
+            updated_func = self.visit_expr(func)
+            updated_func = remove_all_unused(updated_func)
+            self.builder_.update_func(gv, updated_func)
+
+        return self.builder_.get()
+    
+    def visit_function_(self, f):
+        body = super().visit_expr(f.body)
+        params = list(f.params) + list(self.new_params)
+        return tvm.relax.Function(params, body, f.ret_struct_info, f.is_pure, f.attrs, f.span)
+
+    def visit_call_(self, call: relax.Call) -> relax.Expr:
+        call = super().visit_call_(call)
+        if call.op != tvm.ir.Op.get("relax.matmul"):
+            return call
+        permute = self.lookup_binding(call.args[1])
+        if permute is None or permute.op != tvm.ir.Op.get("relax.permute_dims"):
+            return call
+        act = call.args[0]
+        weights = permute.args[0]
+        if weights.struct_info.ndim != 2:
+            return call
+        if act.struct_info.ndim != 2 and act.struct_info.ndim != 3:
+            return call
+
+        def make_scale_param(shape: relax.ShapeExpr, dtype: str) -> tvm.relax.Var:
+            n = 1 if self.mode == "quantize" else shape[-1]
+            scale = relax.Var(f"sq_scale_{self.sm_counter}", relax.TensorStructInfo([n], dtype))
+            self.sm_counter += 1
+            self.new_params.append(scale)
+            return scale
+
+        a_scale = make_scale_param(act.struct_info.shape, act.struct_info.dtype)
+        w_scale = make_scale_param(weights.struct_info.shape, weights.struct_info.dtype)
+        lhs = R.smooth(act, a_scale, kind=1, mode="identity")
+        rhs = R.smooth(weights, w_scale, kind=2, mode="identity")
+        return R.linear(lhs, rhs)
```

Comment:
Since this pass is always followed by `SmoothQuantStatCollector`, which replaces this dummy `smooth` with another dummy op, why not add absmax here instead of adding the first dummy `smooth`?

I think that will let us get rid of the need for the `identity` mode, which in turn might also remove the need for `R.smooth` altogether (instead we can just use `R.multiply`).

Reply:
hmm... I need to think about it but I got your point and I like your idea to remove `R.smooth`!

Also I use `R.smooth` op to lower it to `qnn.quantize` during calibration. Instead I can add `qnn.quantize` op. Then we can remove `R.smooth` if I am not mistaken. What do you think about it?

Reply:
yeah, rather than adding `R.smooth`, I think we should just go ahead and add `quantize / dequantize`. We don't need a full-blown "QNN dialect", we just need these two ops.